File: swift-bridge/Sources/AVAudioBridge/UnitDelay.swift
import AVFoundation
import Foundation

@_cdecl("av_audio_unit_delay_create")
public func av_audio_unit_delay_create() -> UnsafeMutableRawPointer? {
    Unmanaged.passRetained(AVAudioUnitDelay()).toOpaque()
}

@_cdecl("av_audio_unit_delay_release")
public func av_audio_unit_delay_release(_ ptr: UnsafeMutableRawPointer?) {
    guard let ptr else { return }
    Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).release()
}

@_cdecl("av_audio_unit_delay_get_delay_time")
public func av_audio_unit_delay_get_delay_time(_ ptr: UnsafeMutableRawPointer) -> Double {
    Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue().delayTime
}

@_cdecl("av_audio_unit_delay_set_delay_time")
public func av_audio_unit_delay_set_delay_time(_ ptr: UnsafeMutableRawPointer, _ delayTime: Double) {
    let node = Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue()
    node.delayTime = delayTime
}

@_cdecl("av_audio_unit_delay_get_feedback")
public func av_audio_unit_delay_get_feedback(_ ptr: UnsafeMutableRawPointer) -> Float {
    Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue().feedback
}

@_cdecl("av_audio_unit_delay_set_feedback")
public func av_audio_unit_delay_set_feedback(_ ptr: UnsafeMutableRawPointer, _ feedback: Float) {
    let node = Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue()
    node.feedback = feedback
}

@_cdecl("av_audio_unit_delay_get_low_pass_cutoff")
public func av_audio_unit_delay_get_low_pass_cutoff(_ ptr: UnsafeMutableRawPointer) -> Float {
    Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue().lowPassCutoff
}

@_cdecl("av_audio_unit_delay_set_low_pass_cutoff")
public func av_audio_unit_delay_set_low_pass_cutoff(_ ptr: UnsafeMutableRawPointer, _ lowPassCutoff: Float) {
    let node = Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue()
    node.lowPassCutoff = lowPassCutoff
}

@_cdecl("av_audio_unit_delay_get_wet_dry_mix")
public func av_audio_unit_delay_get_wet_dry_mix(_ ptr: UnsafeMutableRawPointer) -> Float {
    Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue().wetDryMix
}

@_cdecl("av_audio_unit_delay_set_wet_dry_mix")
public func av_audio_unit_delay_set_wet_dry_mix(_ ptr: UnsafeMutableRawPointer, _ wetDryMix: Float) {
    let node = Unmanaged<AVAudioUnitDelay>.fromOpaque(ptr).takeUnretainedValue()
    node.wetDryMix = wetDryMix
}

File: src/unit_delay.rs
//! [`AudioUnitDelay`] — feedback delay processing.

#![allow(
    clippy::missing_errors_doc,
    clippy::module_name_repetitions,
    clippy::must_use_candidate
)]

use core::ffi::c_void;
use core::ptr;

use crate::error::AVAudioError;
use crate::ffi;
use crate::node::AudioNodeHandle;
use crate::unit_effect::AudioUnitHandle;

/// Wraps an `AVAudioUnitDelay` node.
pub struct AudioUnitDelay {
    pub(crate) ptr: *mut c_void,
}

unsafe impl Send for AudioUnitDelay {}

impl Drop for AudioUnitDelay {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::av_audio_unit_delay_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

impl AudioNodeHandle for AudioUnitDelay {
    fn as_node_ptr(&self) -> *mut c_void {
        self.ptr
    }
}

impl AudioUnitHandle for AudioUnitDelay {
    fn as_audio_unit_ptr(&self) -> *mut c_void {
        self.ptr
    }
}

impl AudioUnitDelay {
    /// Creates a delay unit.
    pub fn new() -> Result<Self, AVAudioError> {
        let ptr = unsafe { ffi::av_audio_unit_delay_create() };
        if ptr.is_null() {
            return Err(AVAudioError::OperationFailed(
                "failed to create AVAudioUnitDelay".into(),
            ));
        }
        Ok(Self { ptr })
    }

    /// Returns the delay time in seconds.
    pub fn delay_time(&self) -> f64 {
        unsafe { ffi::av_audio_unit_delay_get_delay_time(self.ptr) }
    }

    /// Sets the delay time in seconds.
    pub fn set_delay_time(&self, delay_time: f64) {
        unsafe { ffi::av_audio_unit_delay_set_delay_time(self.ptr, delay_time) };
    }

    /// Returns the feedback percentage.
    pub fn feedback(&self) -> f32 {
        unsafe { ffi::av_audio_unit_delay_get_feedback(self.ptr) }
    }

    /// Sets the feedback percentage.
    pub fn set_feedback(&self, feedback: f32) {
        unsafe { ffi::av_audio_unit_delay_set_feedback(self.ptr, feedback) };
    }

    /// Returns the low-pass cutoff frequency in hertz.
    pub fn low_pass_cutoff(&self) -> f32 {
        unsafe { ffi::av_audio_unit_delay_get_low_pass_cutoff(self.ptr) }
    }

    /// Sets the low-pass cutoff frequency in hertz.
    pub fn set_low_pass_cutoff(&self, low_pass_cutoff: f32) {
        unsafe { ffi::av_audio_unit_delay_set_low_pass_cutoff(self.ptr, low_pass_cutoff) };
    }

    /// Returns the wet/dry mix percentage.
    pub fn wet_dry_mix(&self) -> f32 {
        unsafe { ffi::av_audio_unit_delay_get_wet_dry_mix(self.ptr) }
    }

    /// Sets the wet/dry mix percentage.
    pub fn set_wet_dry_mix(&self, wet_dry_mix: f32) {
        unsafe { ffi::av_audio_unit_delay_set_wet_dry_mix(self.ptr, wet_dry_mix) };
    }
}
